### ISSUE TITLE
Bug when OpenMOC Tries to Handle Non-Square Lattices

### ISSUE THREAD

**Issue by geogunow** [mit-crpg/OpenMOC#135]
When creating a ``Lattice`` with a differing number of cells in the x and y directions, OpenMOC encounters a segmentation fault. 

For instance, if I wish to fill a lattice with the same universe in a rectangular configuration I might initialize a lattice as:
```python
lattice = Lattice(name="lattice")
lattice.setWidth(pitch, pitch)
template = [[universeA]*numcells_x]*numcells_y
lattice.setUniverses(template)
```

If numcells_x (number of cells along the x direction) equals numcells_y (number of cells along the y direction), then the code executes as expected. However, if these numbers differ, then the segmentation fault is encountered.

Sample code is provided below that demonstrates the behavior:

```python
import numpy
from openmoc import *
import openmoc.log as log
import openmoc.plotter as plotter
from openmoc.options import Options


###############################################################################
#######################   Main Simulation Parameters   ########################
###############################################################################

options = Options()

num_threads = options.getNumThreads()
track_spacing = 0.1
num_azim = 32
tolerance = options.getTolerance()
max_iters = options.getMaxIterations()

log.set_log_level('NORMAL')

log.py_printf('TITLE', 'Simulating a two group homogeneous infinite medium...')
log.py_printf('HEADER', 'The reference keff = 1.72...')


###############################################################################
###########################   Creating Materials   ############################
###############################################################################

log.py_printf('NORMAL', 'Creating materials...')

sigT = [1]
sigA = [0.5]
nuSigF = [0.5]
chi = [1]
sigS = [[0.5]]

infinite_medium = Material(name="mat")
infinite_medium.setNumEnergyGroups(1)
infinite_medium.setSigmaA(numpy.array([0.5]))
infinite_medium.setSigmaF(numpy.array([0.5]))
infinite_medium.setNuSigmaF(numpy.array([0.5]))
infinite_medium.setSigmaS(numpy.array([0.5]))
infinite_medium.setChi(numpy.array([1.0]))
infinite_medium.setSigmaT(numpy.array([1.0]))

###############################################################################
###########################   Creating Surfaces   #############################
###############################################################################

log.py_printf('NORMAL', 'Creating surfaces...')

pitch = 30.0
left = XPlane(x=-pitch/2)
right = XPlane(x=pitch/2)
top = YPlane(y=pitch/2)
bottom = YPlane(y=-pitch/2)

left.setBoundaryType(REFLECTIVE)
right.setBoundaryType(REFLECTIVE)
top.setBoundaryType(VACUUM)
bottom.setBoundaryType(REFLECTIVE)


###############################################################################
#############################   Creating Cells   ##############################
###############################################################################

log.py_printf('NORMAL', 'Creating cells...')

cells = []
cells.append(CellBasic(name="fuel"))
cells[0].setMaterial(infinite_medium)

cells.append(CellFill(name="root_cell"))
universeA = Universe(name="universe")

root_universe = Universe(name="root")
universeA.addCell(cells[0])
root_universe.addCell(cells[1])

cells[1].addSurface(halfspace=+1, surface=left)
cells[1].addSurface(halfspace=-1, surface=right)
cells[1].addSurface(halfspace=+1, surface=bottom)
cells[1].addSurface(halfspace=-1, surface=top)


###############################################################################
###########################   Creating Lattices   #############################
###############################################################################

log.py_printf('NORMAL', 'Creating simple pin cell lattice...')

numcells = 2
numcellsx = 10
numcellsy = 50
lattice = Lattice(name="lattice")
lattice.setWidth(pitch, pitch)
template = [[universeA]*numcellsx]*numcellsy
lattice.setUniverses(template)

cells[1].setFill(lattice)

###############################################################################
##########################   Creating the Geometry   ##########################
###############################################################################

log.py_printf('NORMAL', 'Creating geometry...')

geometry = Geometry()
geometry.setRootUniverse(root_universe)
geometry.initializeFlatSourceRegions()

###############################################################################
########################   Creating the TrackGenerator   ######################
###############################################################################

log.py_printf('NORMAL', 'Initializing the track generator...')

track_generator = TrackGenerator(geometry, num_azim, track_spacing)
track_generator.generateTracks()


###############################################################################
###########################   Running a Simulation   ##########################
###############################################################################

solver = CPUSolver(geometry, track_generator)
solver.setNumThreads(num_threads)
solver.setSourceConvergenceThreshold(tolerance)
solver.convergeSource(max_iters)
solver.printTimerReport()

log.py_printf('TITLE', 'Finished')
```

**Comment by wbinventor** [comment]
I just created PR #136 which resolves this issue. I might point out that your input file actually has a bug in it - you should set the widths of your ``Lattice`` cells as follows:

```python
lattice.setWidth(pitch / numcellsx, pitch / numcellsy)
```

If you instead set the width to the ``pitch`` for both x and y as before, then only a small portion of your ``Lattice`` actually resides within the ``XPlane`` and ``YPlane`` objects that you are using as bounds for the geometry.

**Comment by wbinventor** [comment]
Also, could you could package up your script and submit it as a new sample input file in a pull request? It would be nice to illustrate how users can create irregularly sized ``Lattice`` objects to easily mesh 1D and 2D slab geometries.

**Comment by samuelshaner** [comment]
@geogunow thanks for point this out and @wbinventor thanks for submitting a PR. The PR #136 has been merged.

@geogunow - could you submit a pull request with a new sample input file that illustrates how users can create irregular sized `Lattice` objects?

**Comment by wbinventor** [comment]
This issue was closed with the merger of PR #136.
